src/Dependencies/Collections/SegmentedHashSet`1.cs

```diff
@@ -0,0 +1,1454 @@
+{
+    [DebuggerTypeProxy(typeof(ICollectionDebugView<>))]
+    [DebuggerDisplay("Count = {Count}")]
+    internal class SegmentedHashSet<T> : ICollection<T>, ISet<T>, IReadOnlyCollection<T>
+#if NET5_0_OR_GREATER
+        , IReadOnlySet<T>
+#endif
+    {
+        private const bool SupportsComparerDevirtualization
+#if NETCOREAPP
+            = true;
+#else
+            = false;
+#endif
+
+        // This uses the same array-based implementation as Dictionary<TKey, TValue>.
+
+        /// <summary>Cutoff point for stackallocs. This corresponds to the number of ints.</summary>
+        private const int StackAllocThreshold = 100;
+
+        /// <summary>
+        /// When constructing a hashset from an existing collection, it may contain duplicates,
+        /// so this is used as the max acceptable excess ratio of capacity to count. Note that
+        /// this is only used on the ctor and not to automatically shrink if the hashset has, e.g,
+        /// a lot of adds followed by removes. Users must explicitly shrink by calling TrimExcess.
+        /// This is set to 3 because capacity is acceptable as 2x rounded up to nearest prime.
+        /// </summary>
+        private const int ShrinkThreshold = 3;
+        private const int StartOfFreeList = -3;
+
+        private SegmentedArray<int> _buckets;
+        private SegmentedArray<Entry> _entries;
+        private ulong _fastModMultiplier;
+        private int _count;
+        private int _freeList;
+        private int _freeCount;
+        private int _version;
+#if NETCOREAPP
+        private readonly IEqualityComparer<T>? _comparer;
+#else
+        /// <summary>
+        /// <see cref="EqualityComparer{T}.Default"/> doesn't devirtualize on .NET Framework, so we always ensure
+        /// <see cref="_comparer"/> is initialized to a non-<see langword="null"/> value.
+        /// </summary>
+        private readonly IEqualityComparer<T> _comparer;
+#endif
+
+        #region Constructors
+
+        public SegmentedHashSet() : this((IEqualityComparer<T>?)null) { }
+
+        public SegmentedHashSet(IEqualityComparer<T>? comparer)
+        {
+            if (comparer != null && comparer != EqualityComparer<T>.Default) // first check for null to avoid forcing default comparer instantiation unnecessarily
+            {
+                _comparer = comparer;
+            }
+
+#if !NETCOREAPP
+            // .NET Framework doesn't support devirtualization, so we always initialize comparer to a non-null value
+            _comparer ??= EqualityComparer<T>.Default;
+#endif
+        }
+
+        public SegmentedHashSet(int capacity) : this(capacity, null) { }
+
+        public SegmentedHashSet(IEnumerable<T> collection) : this(collection, null) { }
+
+        public SegmentedHashSet(IEnumerable<T> collection, IEqualityComparer<T>? comparer) : this(comparer)
+        {
+            if (collection == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.collection);
```

Comment:
Why this is not `nameof(collection)` (applies to all `ExceptionArgument` usages)

Reply:
➡️ This implementation is taken from the link at the top of the file, and only changed where necessary to build in Roslyn. The original code for this method does not use `nameof(collection)`, so this question is probably better for dotnet/runtime.